File: frontend/lib/jsonl.ts
export type JsonlRecord = Record<string, unknown>

export type JsonlValidationResult =
  | { ok: true; records: JsonlRecord[]; lineCount: number }
  | { ok: false; error: string }

export function validateJsonl(raw: string): JsonlValidationResult {
  const lines = raw.split("\n").map((line) => line.trim()).filter(Boolean)
  if (!lines.length) {
    return { ok: false, error: "JSONL file is empty" }
  }

  const records: JsonlRecord[] = []
  for (let i = 0; i < lines.length; i++) {
    try {
      const row = JSON.parse(lines[i]) as unknown
      if (!row || typeof row !== "object" || Array.isArray(row)) {
        return { ok: false, error: `Line ${i + 1} must be a JSON object` }
      }
      if (!isValidRecord(row as JsonlRecord)) {
        return {
          ok: false,
          error: `Line ${i + 1} needs question+answer, title+content, or text`,
        }
      }
      records.push(row as JsonlRecord)
    } catch {
      return { ok: false, error: `Invalid JSON on line ${i + 1}` }
    }
  }

  return { ok: true, records, lineCount: lines.length }
}

function isValidRecord(row: JsonlRecord): boolean {
  const question = str(row.question ?? row.q)
  const answer = str(row.answer ?? row.a)
  const title = str(row.title ?? row.topic)
  const content = str(row.content ?? row.body)
  const text = str(row.text)

  if (question && answer) return true
  if (title && content) return true
  if (text) return true
  if (question || content) return true
  return false
}

function str(value: unknown): string {
  return typeof value === "string" ? value.trim() : ""
}

export const SAMPLE_JSONL = `{"question":"What is Perai?","answer":"Perai helps companies deploy AI chat widgets trained on their knowledge base."}
{"title":"Pricing","content":"Starter $29/month, Business $99/month, Enterprise contact sales."}
{"text":"Support email: support@perai.io"}
`

export const JSONL_FORMAT_HELP = [
  "One JSON object per line",
  '{"question":"...","answer":"..."}',
  '{"title":"...","content":"..."}',
  '{"text":"..."}',
]

File: backend/app/utils/jsonl.py
import json

MAX_JSONL_BYTES = 10 * 1024 * 1024
MAX_JSONL_LINES = 5000


def parse_jsonl_records(raw: str) -> list[dict]:
    if len(raw.encode("utf-8")) > MAX_JSONL_BYTES:
        raise ValueError("JSONL content exceeds 10MB limit")

    lines = [line.strip() for line in raw.splitlines() if line.strip()]
    if not lines:
        raise ValueError("JSONL file is empty")
    if len(lines) > MAX_JSONL_LINES:
        raise ValueError(f"JSONL exceeds {MAX_JSONL_LINES} lines")

    records: list[dict] = []
    for index, line in enumerate(lines, start=1):
        try:
            row = json.loads(line)
        except json.JSONDecodeError as exc:
            raise ValueError(f"Invalid JSON on line {index}") from exc
        if not isinstance(row, dict):
            raise ValueError(f"Line {index} must be a JSON object")
        records.append(row)
    return records


def record_to_block(row: dict, index: int) -> str:
    question = str(row.get("question") or row.get("q") or "").strip()
    answer = str(row.get("answer") or row.get("a") or "").strip()
    title = str(row.get("title") or row.get("topic") or "").strip()
    content = str(row.get("content") or row.get("body") or "").strip()
    text = str(row.get("text") or "").strip()

    if question and answer:
        return f"Question: {question}\nAnswer: {answer}"
    if title and content:
        return f"Title: {title}\nContent: {content}"
    if text:
        return f"Text: {text}"
    if question:
        return f"Question: {question}"
    if content:
        return f"Content: {content}"

    raise ValueError(
        f"Line {index} needs question+answer, title+content, or text"
    )


def jsonl_to_knowledge(raw: str) -> str:
    records = parse_jsonl_records(raw)
    blocks = [record_to_block(row, i) for i, row in enumerate(records, start=1)]
    return "\n\n---\n\n".join(blocks)


def normalize_jsonl_upload(raw: str) -> str:
    return jsonl_to_knowledge(raw)
